### PR TITLE
Add an endpoint for admin user registration

### PR DESCRIPTION
**Pull request by nunemel**
# PR Details
Add an endpoint for admin user registration 

## Description
-Add validation to user login, registration and admin registration end points 
-User registrations

## Related Issue
#1  #4
Closes #44 issue

### CODE REVIEW

File: src/controllers/auth-controller.js [hunk 1]
```diff
@@ -1,38 +1,41 @@
...
-      return res.status(404).json({
-        error: 'User not found'
-      });
+      return Utils.sendErrorResponse(res, notFound, userNotFoundError);
     }
 
-    const token = await JWT.sign({ _id: user._id, email: user.email, role: user.role }, jwtSecret, jwtOptions);
-
-    return res.status(200).json({
+    const token = await TokenLib.createUserToken(user._id, email, user.role);
```
Comment (anahit42) [inline src/controllers/auth-controller.js:30]: pass with object

File: src/controllers/auth-controller.js [hunk 2]
```diff
@@ -1,38 +1,41 @@
...
   try {
     const { email, password } = req.body;
-
     const user = await UserModel.findOne({ email });
 
     if (!user) {
-      return res.status(404).json({
-        error: 'User not found'
-      });
+      return Utils.sendErrorResponse(res, notFound, userNotFoundError);
```
Comment (anahit42) [inline src/controllers/auth-controller.js:21]: throw custom error

File: src/controllers/auth-controller.js [hunk 3]
```diff
@@ -1,38 +1,41 @@
...
-const jwtOptions = config.get('jwt.options');
+const HttpStatus = require('http-status-codes');
 
 const { UserModel } = require('../models');
 const CryptoLib = require('../libs/crypto-lib');
+const TokenLib = require('../libs/token-lib');
+const Utils = require('../utils');
+const AdminRole = require('../constants').Admin;
+
+const notFound = HttpStatus.NOT_FOUND;
```
Comment (anahit42) [inline src/controllers/auth-controller.js:9]: destruct NOT_FOUND, CONFLICT

File: src/controllers/auth-controller.js [hunk 4]
```diff
@@ -1,38 +1,41 @@
...
-      });
+      return Utils.sendErrorResponse(res, notFound, userNotFoundError);
     }
 
-    const token = await JWT.sign({ _id: user._id, email: user.email, role: user.role }, jwtSecret, jwtOptions);
-
-    return res.status(200).json({
+    const token = await TokenLib.createUserToken(user._id, email, user.role);
+    // FOR TESTING ONLY
+    const adminToken = await TokenLib.createAdminToken('babygearadmin');
```
Comment (anahit42) [inline src/controllers/auth-controller.js:32]: remove adminToken

File: src/controllers/auth-controller.js [hunk 5]
```diff
@@ -41,24 +44,87 @@ async function login (req, res, next) {
...
+    return next(error);
+  }
+}
+
+async function registerAdmin(req, res, next) {
+  try {
+    const { email } = req.body;
+    const user = await checkIfUserExists(email);
+    if (user) {
+      return Utils.sendErrorResponse(res, conflict, duplicateUserError);
```
Comment (anahit42) [inline src/controllers/auth-controller.js:85]: throw custom error

File: src/controllers/auth-controller.js [hunk 6]
```diff
@@ -41,24 +44,87 @@ async function login (req, res, next) {
...
+    }
+
+    const { authorization } = req.headers;
+
+    await TokenLib.checkAdminToken(res, authorization);
```
Comment (anahit42) [inline src/controllers/auth-controller.js:90]: remove from here

File: src/controllers/auth-controller.js [hunk 7]
```diff
@@ -41,24 +44,87 @@ async function login (req, res, next) {
...
+
+    const { firstName,
+      lastName,
+      password,
+      mobilePhone,
+      address} = req.body;
```
Comment (anahit42) [inline src/controllers/auth-controller.js:96]: apply styling

File: src/controllers/user-controller.js [hunk 8]
```diff
@@ -1,36 +1,116 @@
...
-const config = require('config');
-
-const jwt = config.get('jwt');
+const HttpStatus = require('http-status-codes');
+const TokenLib = require('../libs/token-lib');
 
 const { UserModel } = require('../models');
 
 async function getUser (req, res, next) {
+
```
Comment (anahit42) [inline src/controllers/user-controller.js:7]: remove blank line

File: src/libs/token-lib.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+async function createUserToken(userId, email, role) {
+  return JWT.sign({ _id: userId, email: email, role: role },
+    jwtSecret,
+    jwtOptions);
+}
+
+// FOR TESTING ONLY
+async function createAdminToken(adminCode) {
+  return JWT.sign({ adminCode: adminCode },
```
Comment (anahit42) [inline src/libs/token-lib.js:17]: apply linting

File: src/libs/token-lib.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    jwtSecret,
+    jwtOptions);
+}
+
+async function checkLoginToken(res, userId, authorization) {
+  const decoded = await JWT.verify(authorization, jwt.secret);
+
+  if (userId !== decoded._id.toString()) {
+    sendErrorResponse(res, HttpStatus.UNAUTHORIZED, HttpStatus.getStatusText(HttpStatus.UNAUTHORIZED));
```
Comment (anahit42) [inline src/libs/token-lib.js:26]: throw custom error

File: src/libs/token-lib.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  }
+}
+
+async function checkAdminToken(res, authorization) {
+  const decoded = await JWT.verify(authorization, jwt.secret);
+
+  if (decoded.adminCode.toString() !== jwt.adminCode) {
+    sendErrorResponse(res, HttpStatus.FORBIDDEN, HttpStatus.getStatusText(HttpStatus.FORBIDDEN));
```
Comment (anahit42) [inline src/libs/token-lib.js:34]: throw custom error

File: src/libs/token-lib.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+}
+
+module.exports = {
+  createUserToken,
+  checkLoginToken,
+  checkAdminToken,
+  createAdminToken };
```
Comment (anahit42) [inline src/libs/token-lib.js:43]: apply linting

File: src/middlewares/validators/schemas/user-schemas.js [hunk 13]
```diff
@@ -1,14 +1,33 @@
...
+
 const registerUserSchema = Joi.object({
   body: Joi.object({
     firstName: Joi.string().trim().max(100).required(),
     lastName: Joi.string().trim().max(100).trim().required(),
+    email: Joi.string().trim().email().required(),
     password: Joi.string().pattern(new RegExp('^[a-zA-Z0-9]{5,30}$')).required(),
-    email: Joi.string().trim().email().required()
+    mobilePhone: Joi.string().pattern(new RegExp('^0\\d{2}-\\d{2}-\\d{2}-\\d{2}$')),
+    //image: Joi.string().pattern(new RegExp('([a-z\\-_0-9\\/\\:\\.]*\\.(jpg|jpeg|png|gif))')),
```
Comment (anahit42) [inline src/middlewares/validators/schemas/user-schemas.js:17]: remove image

File: src/middlewares/validators/user-validator.js [hunk 14]
```diff
@@ -3,22 +3,40 @@ const config = require('config');
 const validationOptions = config.get('validation.options');
 
 const { ValidationError } = require('../../errors');
-const { UserSchemas } = require('./schemas');
+const { UserSchemas,  } = require('./schemas');
```
Comment (anahit42) [inline src/middlewares/validators/user-validator.js:6]: remove comma

File: src/middlewares/validators/user-validator.js [hunk 15]
```diff
@@ -3,22 +3,40 @@ const config = require('config');
...
+  }
+
+  return next();
+}
+
+function validateLoginUser (req, res, next) {
+  const { error } = UserSchemas.loginUserSchema.validate(req, validationOptions);
+
+  if (error) {
+    printErrorDetails(error, next);
```
Comment (anahit42) [inline src/middlewares/validators/user-validator.js:22]: handelError

File: src/middlewares/validators/user-validator.js [hunk 16]
```diff
@@ -3,22 +3,40 @@ const config = require('config');
...
+  }
+
+  return next();
+}
+
+function printErrorDetails(error, next) {
   if (error) {
```
Comment (anahit42) [inline src/middlewares/validators/user-validator.js:29]: remove error check

File: src/middlewares/validators/user-validator.js (hunk 15 above)
Comment (anahit42) [inline src/middlewares/validators/user-validator.js:22]: return from if block

Comment (nunemel) [reply src/middlewares/validators/user-validator.js:22]: it already returns

File: src/middlewares/validators/user-validator.js (hunk 16 above)
Comment (nunemel) [reply src/middlewares/validators/user-validator.js:29]: removed

File: src/middlewares/validators/user-validator.js (hunk 15 above)
Comment (nunemel) [reply src/middlewares/validators/user-validator.js:22]: changed the name to handleErrorDetails

File: src/middlewares/validators/user-validator.js (hunk 14 above)
Comment (nunemel) [reply src/middlewares/validators/user-validator.js:6]: done

File: src/middlewares/validators/schemas/user-schemas.js (hunk 13 above)
Comment (nunemel) [reply src/middlewares/validators/schemas/user-schemas.js:17]: done

File: src/libs/token-lib.js (hunk 12 above)
Comment (nunemel) [reply src/libs/token-lib.js:43]: done

File: src/libs/token-lib.js (hunk 10 above)
Comment (nunemel) [reply src/libs/token-lib.js:26]: done

File: src/libs/token-lib.js (hunk 9 above)
Comment (nunemel) [reply src/libs/token-lib.js:17]: done

File: src/controllers/user-controller.js (hunk 8 above)
Comment (nunemel) [reply src/controllers/user-controller.js:7]: done

File: src/controllers/auth-controller.js (hunk 7 above)
Comment (nunemel) [reply src/controllers/auth-controller.js:96]: դոնե Ճ

File: src/controllers/auth-controller.js (hunk 4 above)
Comment (nunemel) [reply src/controllers/auth-controller.js:32]: done

File: src/controllers/auth-controller.js (hunk 3 above)
Comment (nunemel) [reply src/controllers/auth-controller.js:9]: done

File: src/controllers/auth-controller.js (hunk 2 above)
Comment (nunemel) [reply src/controllers/auth-controller.js:21]: done

File: src/controllers/auth-controller.js (hunk 1 above)
Comment (nunemel) [reply src/controllers/auth-controller.js:30]: done

**Comment by anahit42** [comment]
Committed in other PR

**Pull request closed without merging**